tests/rptest/services/keycloak.py

```diff
@@ -0,0 +1,266 @@
+            'path': f"{KC_LOG_FILE}",
+            "collect_default": True,
+        },
+    }
+
+    def __init__(self, context,
+                 port=KC_PORT,
+                 realm=DEFAULT_REALM,
+                 log_level=KC_ROOT_LOG_LEVEL):
+        super(KeycloakService, self).__init__(context, num_nodes=1)
+        self.realm = realm
+        self.http_port = port
+        self.log_level = log_level
+
+    def _start_cmd(self):
+        cmd = START_CMD_TMPL.format(
+            admin=KC_ADMIN,
+            pw=KC_ADMIN_PASSWORD,
+            kc=KC, port=self.http_port,
+            log_handler=KC_LOG_HANDLER, logfile=KC_LOG_FILE,
+            log_level=self.log_level)
+        return cmd
+
+    def _construct_admin_cmd(self, args: list[str]):
+        cmd = f"{KCADM} {' '.join(args)}"
+        return cmd
+
+    # Always in the master realm, where the admin user lives
+    def _config_creds(self, node, realm='master', user=KC_ADMIN, password=KC_ADMIN_PASSWORD):
+        cmd = self._construct_admin_cmd([
+            'config credentials',
+            f'--server http://localhost:{self.http_port}',
+            f'--realm {realm}',
+            f'--user {user}',
+            f'--password {password}',
+        ])
+        node.account.ssh(cmd, allow_fail=False)
+
+    def _create_realm(self, node):
+        rep = {
+            'realm': self.realm,
+            'enabled': True,
+        }
+        cmd = self._construct_admin_cmd([
+            'create realms',
+            '-f -',
+            f"<<< '{json.dumps(rep)}'"
+        ])
+        node.account.ssh(cmd, allow_fail=False)
+
+    def _ssh_decode_json(self, node, cmd, verbose=False, strict=True):
+        bin_out = node.account.ssh_output(cmd, allow_fail=not strict)
+        if verbose:
+            self.logger.info(f"{bin_out}")
+        return json.loads(bin_out.decode('utf-8'))
+
+    def create_client(self, node, client_id, **kwargs):
+        rep = {
+            'clientId': client_id,
+            'enabled': True,
+            'serviceAccountsEnabled': True, # for client credentials grant
+            # 'standardFlowEnabled': True,  # authorization flow grant
+            # 'directAccessGrantsEnabled': True, # resource owner password grant
+            # 'implicitFlowEnabled': True,  # implicit grant
+        }
+        rep.update(kwargs)
+        cmd = self._construct_admin_cmd([
+            'create clients',
+            f'-r {self.realm}',
+            '-f -',
+            f"<<< '{json.dumps(rep)}'"
+        ])
+        out = node.account.ssh_output(cmd, allow_fail=False)
+        self.logger.debug(out)
+
+    def ip(self, node):
+        return node.account.externally_routable_ip
+
+    def get_clients(self, node, realm, client_id=None):
+        args = ['get', 'clients', f'-r {realm}']
+        if client_id is not None:
+            args.append(f'--query clientId={client_id}')
+
+        cmd = self._construct_admin_cmd(args)
+        return self._ssh_decode_json(node, cmd)
```

Comment:
pretty neat. so this is using some keycloak cli tooling?

Reply:
not anymore  😅